nemo_rl/experience/rollouts.py

```diff
@@ -388,9 +483,389 @@ def run_multi_turn_rollout(
+    max_seq_len: int,
+    max_rollout_turns: int = 999999,
+    greedy: bool = False,
+) -> tuple[dict, dict[str, Any]]:
+    """Run a multi-turn rollout for a single sample.
+
+    This function manages the complete lifecycle of one sample's interaction.
+    Async generation is used internally when available.
+
+    Args:
+        sample_idx: Index of this sample in the original batch
+        initial_sample_state: Initial state containing message_log, extra_env_info, etc.
+        policy_generation: The generation interface
+        tokenizer: Tokenizer to use
+        task_to_env: Environment mapping
+        max_seq_len: Maximum sequence length
+        max_rollout_turns: Maximum number of turns
+        greedy: Whether to use greedy decoding
+
+    Returns:
+        Tuple of (final_sample_state, sample_metrics)
+    """
+    # Initialize sample state
+    current_message_log = copy.deepcopy(initial_sample_state["message_log"])
+    current_extra_env_info = copy.deepcopy(initial_sample_state["extra_env_info"])
+    current_stop_strings = initial_sample_state.get("stop_strings", None)
+    task_name = initial_sample_state["task_name"]
+
+    # Sample-level metrics
+    total_reward = 0.0
+    turn_count = 0
+    token_count = 0
+    assistant_token_count = 0
+    env_token_count = 0
+    terminated = False
+    truncated = False
+    max_turns_reached = False
+
+    # Track per-turn metrics
+    turn_gen_tokens = []
+
+    for turn in range(max_rollout_turns):
+        if terminated or truncated:
+            break
+
+        turn_count += 1
+
+        # Generate response for this sample using async generation
+        try:
+            (
+                updated_message_log,
+                generated_tokens,
+                input_lengths,
+                gen_metrics,
+            ) = await async_generate_response_for_sample_turn(
+                policy_generation,
+                current_message_log,
+                current_stop_strings,
+                tokenizer,
+                max_seq_len,
+                greedy=greedy,
+            )
+            current_message_log = updated_message_log
+
+            # Update token counts
+            gen_token_count = len(generated_tokens)
+            assistant_token_count += gen_token_count
+            token_count += gen_token_count
+            turn_gen_tokens.append(gen_token_count)
+
+        except Exception as e:
+            print(f"Error generating response for sample {sample_idx}: {e}")
+            break
+
+        # Create single-sample batch for environment interaction
+        sample_batch = BatchedDataDict[DatumSpec](
+            {
+                "message_log": [current_message_log],
+                "extra_env_info": [current_extra_env_info],
+                "task_name": [task_name],
+            }
+        )
+
+        # Get environment feedback
+        try:
+            env_output = calculate_rewards(sample_batch, task_to_env)
```

Comment:
I assume the calculate rewards is still a single process single-thread? Would all requests to this be processed one-by-one?

Reply:
Yes, will be addressing this in a separate PR.